---
title: "Annotating gnomad chr21 for escape from NMD"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{gnomad_chr21}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

```

## Download and parse gnomad chromosome 21

```{r read_data, message=FALSE, eval=TRUE}
library(aenmd)
library(BiocFileCache)

vcf_url <- 'https://storage.googleapis.com/gcp-public-data--gnomad/release/2.1.1/liftover_grch38/vcf/exomes/gnomad.exomes.r2.1.1.sites.21.liftover_grch38.vcf.bgz'
vcf <- aenmd:::parse_vcf_vcfR(bfcrpath(BiocFileCache(), vcf_url))
```

Next we process variants, focusing keeping only those that 

* Overlap tsl=1 transcripts (Ensembl v. 105)
* Don't overlap splice regions
* Are contained in at least one exon
* If the variants are SNVs, they do create a PTC

```{r, process_data, eval=TRUE}
var_rng <- process_variants(vcf$vcf_rng, verbose = TRUE)
var_rng |> print()
```
Finally, we annotate these variants for escape from NMD.

```{r, annotate_variants, eval=TRUE}
#- remove the substetting below for complete chr21
res <- annotate_nmd(var_rng[1:50])
res |> print()
```
...and we can focus on NMD-escaping variants:

```{r, focus, eval= FALSE}
ind <- S4Vectors::mcols(res)$res_aenmd[,"is_ptc"] & (rowSums(S4Vectors::mcols(res)$res_aenmd[,1:6] |> as.matrix()) >1 ) 
res[ind]
table(res[ind]$type)
```
